File: tools/pineconnector/python/queue.py
"""ZeroMQ wrappers for signal dispatch and result consumption."""

from __future__ import annotations

import asyncio
import logging
from typing import Optional

import zmq
import zmq.asyncio

from .models import ValidatedSignal

log = logging.getLogger(__name__)


class ZMQProducer:
    """PUSH socket — sends validated signals to trade engine."""

    def __init__(self, address: str) -> None:
        self.address = address
        self._ctx = zmq.asyncio.Context.instance()
        self._socket = self._ctx.socket(zmq.PUSH)
        self._socket.setsockopt(zmq.SNDHWM, 1000)
        self._socket.setsockopt(zmq.LINGER, 1000)
        self._socket.connect(address)
        self.connected = True
        log.info("ZMQ producer connected to %s", address)

    async def send(self, signal: ValidatedSignal) -> None:
        try:
            await self._socket.send_json(signal.model_dump())
        except zmq.ZMQError as e:
            self.connected = False
            log.error("ZMQ send error: %s", e)
            raise

    def close(self) -> None:
        self._socket.close()
        self.connected = False


class ZMQConsumer:
    """PULL socket — receives execution results from MT5 bridge."""

    def __init__(self, address: str) -> None:
        self.address = address
        self._ctx = zmq.asyncio.Context.instance()
        self._socket = self._ctx.socket(zmq.PULL)
        self._socket.setsockopt(zmq.RCVHWM, 1000)
        self._socket.setsockopt(zmq.LINGER, 1000)
        self._socket.bind(address)
        log.info("ZMQ result consumer bound to %s", address)

    async def receive(self) -> Optional[dict]:
        try:
            if await self._socket.poll(timeout=100):
                return await self._socket.recv_json()
            return None
        except zmq.ZMQError as e:
            log.error("ZMQ recv error: %s", e)
            return None

    def close(self) -> None:
        self._socket.close()


class ZMQStateSubscriber:
    """SUB socket — receives state updates from trade engine."""

    def __init__(self, address: str) -> None:
        self.address = address
        self._ctx = zmq.asyncio.Context.instance()
        self._socket = self._ctx.socket(zmq.SUB)
        self._socket.setsockopt(zmq.RCVHWM, 1000)
        self._socket.setsockopt(zmq.LINGER, 1000)
        self._socket.setsockopt_string(zmq.SUBSCRIBE, "")
        self._socket.connect(address)
        log.info("ZMQ state subscriber connected to %s", address)

    async def receive(self) -> Optional[dict]:
        try:
            if await self._socket.poll(timeout=100):
                return await self._socket.recv_json()
            return None
        except zmq.ZMQError as e:
            log.error("ZMQ state recv error: %s", e)
            return None

    def close(self) -> None:
        self._socket.close()

File: tools/pineconnector/rust/src/queue.rs
use std::thread;

use anyhow::Result;
use tokio::sync::mpsc;
use tracing::{error, info, warn};

use crate::models::{ExecutionCommand, ExecutionResult, StateUpdate, ValidatedSignal};

/// Spawn an OS thread that binds a ZMQ PULL socket and forwards
/// deserialized signals into a tokio mpsc channel.
pub fn start_signal_receiver(
    address: &str,
    tx: mpsc::UnboundedSender<ValidatedSignal>,
) -> thread::JoinHandle<()> {
    let addr = address.to_string();
    thread::Builder::new()
        .name("zmq-signal-rx".into())
        .spawn(move || {
            let ctx = zmq::Context::new();
            let socket = ctx.socket(zmq::PULL).expect("Failed to create PULL socket");
            socket.set_rcvhwm(1000).ok();
            socket.bind(&addr).unwrap_or_else(|e| {
                panic!("Failed to bind PULL socket on {addr}: {e}");
            });
            info!("Signal receiver bound on {}", addr);

            loop {
                match socket.recv_msg(0) {
                    Ok(msg) => {
                        let bytes = msg.as_ref();
                        match serde_json::from_slice::<ValidatedSignal>(bytes) {
                            Ok(signal) => {
                                if tx.send(signal).is_err() {
                                    warn!("Signal channel closed, exiting receiver");
                                    break;
                                }
                            }
                            Err(e) => {
                                error!(
                                    "Failed to deserialize signal: {} | raw: {}",
                                    e,
                                    String::from_utf8_lossy(bytes)
                                );
                            }
                        }
                    }
                    Err(e) => {
                        error!("ZMQ recv error: {}", e);
                        thread::sleep(std::time::Duration::from_secs(1));
                    }
                }
            }
        })
        .expect("Failed to spawn signal receiver thread")
}

/// Spawn an OS thread that connects a ZMQ PUSH socket and sends
/// execution commands received from a tokio mpsc channel.
pub fn start_command_sender(
    address: &str,
    mut rx: mpsc::UnboundedReceiver<ExecutionCommand>,
) -> thread::JoinHandle<()> {
    let addr = address.to_string();
    thread::Builder::new()
        .name("zmq-cmd-tx".into())
        .spawn(move || {
            let ctx = zmq::Context::new();
            let socket = ctx.socket(zmq::PUSH).expect("Failed to create PUSH socket");
            socket.set_sndhwm(1000).ok();
            socket.set_linger(1000).ok();
            socket.connect(&addr).unwrap_or_else(|e| {
                panic!("Failed to connect PUSH socket to {addr}: {e}");
            });
            info!("Command sender connected to {}", addr);

            // Block on the mpsc receiver using a blocking runtime bridge
            let rt = tokio::runtime::Builder::new_current_thread()
                .enable_all()
                .build()
                .expect("Failed to build mini runtime");

            rt.block_on(async {
                while let Some(cmd) = rx.recv().await {
                    match serde_json::to_vec(&cmd) {
                        Ok(data) => {
                            if let Err(e) = socket.send(&data, 0) {
                                error!("ZMQ send error for cmd {}: {}", cmd.command_id, e);
                            }
                        }
                        Err(e) => {
                            error!("Failed to serialize command: {}", e);
                        }
                    }
                }
            });

            warn!("Command channel closed, exiting sender");
        })
        .expect("Failed to spawn command sender thread")
}

/// Spawn an OS thread that binds a ZMQ PUB socket and publishes
/// state updates received from a tokio mpsc channel.
pub fn start_state_publisher(
    address: &str,
    mut rx: mpsc::UnboundedReceiver<StateUpdate>,
) -> thread::JoinHandle<()> {
    let addr = address.to_string();
    thread::Builder::new()
        .name("zmq-state-pub".into())
        .spawn(move || {
            let ctx = zmq::Context::new();
            let socket = ctx.socket(zmq::PUB).expect("Failed to create PUB socket");
            socket.set_sndhwm(1000).ok();
            socket.set_linger(1000).ok();
            socket.bind(&addr).unwrap_or_else(|e| {
                panic!("Failed to bind PUB socket on {addr}: {e}");
            });
            info!("State publisher bound on {}", addr);

            let rt = tokio::runtime::Builder::new_current_thread()
                .enable_all()
                .build()
                .expect("Failed to build mini runtime");

            rt.block_on(async {
                while let Some(update) = rx.recv().await {
                    match serde_json::to_vec(&update) {
                        Ok(data) => {
                            if let Err(e) = socket.send(&data, 0) {
                                error!("ZMQ pub error: {}", e);
                            }
                        }
                        Err(e) => {
                            error!("Failed to serialize state update: {}", e);
                        }
                    }
                }
            });

            warn!("State channel closed, exiting publisher");
        })
        .expect("Failed to spawn state publisher thread")
}

/// Spawn an OS thread that connects a ZMQ PULL socket to receive
/// execution results from the MT5 bridge, forwarding them into a channel.
pub fn start_result_receiver(
    address: &str,
    tx: mpsc::UnboundedSender<ExecutionResult>,
) -> thread::JoinHandle<()> {
    let addr = address.to_string();
    thread::Builder::new()
        .name("zmq-result-rx".into())
        .spawn(move || {
            let ctx = zmq::Context::new();
            let socket = ctx.socket(zmq::PULL).expect("Failed to create result PULL socket");
            socket.set_rcvhwm(1000).ok();
            // Connect (not bind) — Python binds on this port for results
            // But Rust also needs results for state management.
            // We use a SUB socket to tap into results published by the bridge.
            // Alternative: the bridge sends results to both Python and Rust.
            // For simplicity, Rust receives results via a separate PULL from bridge.
            socket.connect(&addr).unwrap_or_else(|e| {
                panic!("Failed to connect result PULL to {addr}: {e}");
            });
            info!("Result receiver connected to {}", addr);

            loop {
                match socket.recv_msg(0) {
                    Ok(msg) => {
                        match serde_json::from_slice::<ExecutionResult>(msg.as_ref()) {
                            Ok(result) => {
                                if tx.send(result).is_err() {
                                    warn!("Result channel closed, exiting");
                                    break;
                                }
                            }
                            Err(e) => {
                                error!("Failed to deserialize result: {}", e);
                            }
                        }
                    }
                    Err(e) => {
                        error!("ZMQ result recv error: {}", e);
                        thread::sleep(std::time::Duration::from_secs(1));
                    }
                }
            }
        })
        .expect("Failed to spawn result receiver thread")
}
